static/js/compose_error.ts

```diff
@@ -1,11 +1,68 @@
 import $ from "jquery";
 
+import render_compose_banner from "../templates/compose_banner.hbs";
+import render_stream_does_not_exist_error from "../templates/stream_does_not_exist_error.hbs";
+
 import * as common from "./common";
 
 // banner types
 export const WARNING = "warning";
+export const ERROR = "error";
+
+const WARNING_MESSAGES = [".topic_resolved"];
+
+const ERROR_MESSAGES = [
+    ".empty_message",
+    ".wildcards_not_allowed",
+    ".subscription_error",
+    ".stream_does_not_exist",
+    ".missing_stream",
+    ".no_post_permissions",
+    ".private_messages_disabled",
+    ".missing_private_message_recipient",
+    ".invalid_recipient",
+    ".invalid_recipients",
+    ".deactivated_user",
+    ".message_too_long",
+    ".topic_missing",
+    ".zephyr_not_running",
+    ".generic_compose_error",
+];
+
+export const ALL_BANNER_MESSAGES = WARNING_MESSAGES.concat(ERROR_MESSAGES);
+
+// TODO: Replace with compose_ui.hide_compose_spinner() when it is converted to ts.
+function hide_compose_spinner(): void {
+    $("#compose-send-button .loader").hide();
+    $("#compose-send-button span").show();
+    $("#compose-send-button").removeClass("disable-btn");
+}
+
+// NOTE: if you add a new classname, make sure it's added to ALL_BANNER_MESSAGES above.
+export function show_error_message(message: string, classname: string, $bad_input?: JQuery): void {
+    if (classname === "generic_compose_error" && $(".generic_compose_error").length) {
+        // Generic errors overwrite each other, so if one already exists we just update the text.
+        $(".generic_compose_error .banner_content").text(message);
```

Comment:
Does this block need to do anything to clear non-text content that may have been inserted by other banner types?

Reply:
Also, what's the thinking behind having this block instead of just rerendering the thing from scratch whenever requested? I'm not sure in what situations having a special case here would improve things.

Reply:
> Does this block need to do anything to clear non-text content that may have been inserted by other banner types?

I was wondering about that, and felt mostly confident that it would override that, but not totally sure.

> Also, what's the thinking behind having this block instead of just rerendering the thing from scratch whenever requested? I'm not sure in what situations having a special case here would improve things.

You're totally right, I was stuck in the pattern I was copying from. This should just clear the existing generic message and continue adding a banner normally like the others. That would be way better and address the first issue